### PR TITLE
Add library improvements to tt_dit

### PR DESCRIPTION
**Pull request by friedrich**
This PR is a collection of tt_dit improvements and fixes made while implementing Flux.1, Motif Image and Qwen Image.

Changes:
- remove sp_axis from Conv2d and check input shape
- fix tt_dit tests
- support sequence padding in PatchEmbed
- add option to return normalized state in CLIP
- add support for attention mask in T5
- move linear chunk helper to linear.py
- support input parallelism in Conv2d
- add append to ModuleList
- allow empty prefix in substate
- add DistributedRMSNorm tensor dimension check
- add support for ColParallelLinear and RowParallelLinear without mesh axis
- make Conv2d initialization clearer and support scalar kernel_size, stride, etc
- improve Conv2d out-of-memory error display
- improve mesh distribution error in Module
- make text embedder optional
- allow for already unfolded input in PatchEmbed
- refactor activation function application
- make from_torch and to_torch more ergonomic and remove unused code
- improve types in substate
- be less verbose in assert_quality

### CODE REVIEW

File: models/experimental/tt_dit/layers/conv2d.py [hunk 1]
```diff
@@ -107,55 +121,111 @@ def __init__(
...
 
     def _prepare_torch_state(self, state: dict[str, torch.Tensor]) -> None:
-        if "bias" in state:
-            state["bias"] = state["bias"].reshape([1, 1, 1, -1])
+        bias = state.pop("bias", None)
+        if bias is not None:
+            (out_dim,) = bias.shape
+            bias = bias.reshape([1, 1, 1, out_dim])
+            bias_zeros = torch.zeros([self.in_mesh_axis_size - 1, 1, 1, out_dim])
+            state["bias"] = torch.cat([bias, bias_zeros])
```
[the remark this answers is not in the document]
Comment (friedrich) [reply models/experimental/tt_dit/layers/conv2d.py:229]: Zero is the default

File: models/experimental/tt_dit/layers/conv2d.py [hunk 2]
```diff
@@ -107,55 +121,111 @@ def __init__(
             sp_axis: Axis to use for sequence parallelism. Currently only used for gather before computation
```
Comment (sadesoyeTT) [inline models/experimental/tt_dit/layers/conv2d.py]: Can we add documentation for in_mesh_axis and out_mesh_axis. They should be able to replace sp_axis hack that I added. sp_axis is just used to keep track of which axis the input is sharded on.

File: models/experimental/tt_dit/layers/conv2d.py [hunk 3]
```diff
@@ -170,7 +240,7 @@ def forward(self, x: ttnn.Tensor, /) -> ttnn.Tensor:
         Data is left in the state of the final compute. The burden is on the next layer to prepare its input as needed.
         TODO: Add support for DP and SP
         """
-        if self.is_sharded_tensor(x):
+        if self.sp_axis is not None and self.is_sharded_tensor(x):
```
Comment (sadesoyeTT) [inline models/experimental/tt_dit/layers/conv2d.py]: Can we replace with in/out_mesh_axis.

Comment (friedrich) [reply models/experimental/tt_dit/layers/conv2d.py]: I looked at the SD3.5 VAE, there is only one place, where sp_axis is not equal to out_mesh_axis and it makes sense to move the all-gather out of conv2d there. So I removed sp_axis and replaced this condition to check for out_mesh_axis.

Note that it would not be possible to replace with in_mesh_axis, since this would make the weights sharded on the input dimension as well.

File: models/experimental/tt_dit/layers/conv2d.py (hunk 2 above)
Comment (friedrich) [reply models/experimental/tt_dit/layers/conv2d.py]: I added documentation on the new arguments as well as information on sharding of the input and output tensors.

**Review by sadesoyeTT — APPROVED** [review]

**Comment by friedrich** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/18944450602

**Review by cglagovichTT — APPROVED** [review]
